typing/parmatch.ml

```diff
@@ -1869,14 +1876,28 @@ let extendable_path path =
     Path.same path Predef.path_unit ||
     Path.same path Predef.path_option)
 
+let warn_fragile_type ty env =
+  let _,_,decl = Ctype.extract_concrete_typedecl env ty in
+  let warn = Warnings.Fragile_match "" in
+  match Warnings.status warn with
+    | Warnings.Always -> true
+    | Warnings.Never -> false
+    | Warnings.Implicit | Warnings.Explicit ->
+        Builtin_attributes.with_warning_attribute
+          decl.Types.type_attributes
+          (fun () -> Warnings.(is_active warn))
+
 let rec collect_paths_from_pat r p = match p.pat_desc with
 | Tpat_construct(_, {cstr_tag=(Cstr_constant _|Cstr_block _|Cstr_unboxed)},ps)
   ->
-    let path =  get_type_path p.pat_type p.pat_env in
-    List.fold_left
-      collect_paths_from_pat
-      (if extendable_path path then add_path path r else r)
-      ps
+    (match get_type_path p.pat_type p.pat_env with
+     | Some path ->
+         List.fold_left
+           collect_paths_from_pat
+           (if extendable_path path && warn_fragile_type p.pat_type p.pat_env
+            then add_path path r else r)
+           ps
+     | None -> r)
```

Comment:
`r` is the accumulator (the list of paths we've seen so far), while `ps` is a list of patterns : the arguments of the constructor, i.e. if `p` is `Foo (a, _, None)` then `ps` is `[Tpat_var "a"; Tpat_any; Tpat_construct "None"]`.
Returning `ps` wouldn't make sense (it wouldn't type), while returning `r` does.

One a similar note one can wonder why the code is not instead:
```ocaml
let r =
  match get_type_path ... with
  | Some p when extendable_path p -> add_path p r
  | _ -> r
in
List.fold_left collect_paths_from_pat r ps
```
Here my guess would be "well, we already know we're in a nonsensical situation, so we might as well not do any extra work". It would be nice if the author could add a comment in the code clarifying that though.

As for the last question "why not error", `get_type_path` was previously erroring but was modified not to anymore (and return an option instead), it would be of little benefit if we failed when looking at the result.